proto/p4/p4runtime.proto

```diff
@@ -356,6 +357,29 @@ message CloneSessionEntry {
   uint64 packet_length_bytes = 4;
 }
 
+//------------------------------------------------------------------------------
+// For writing and reading values in a parser value set. It is OK for the value
+// set to have no entries, as a result of a write. A state transition matching
+// on an empty value set will never be taken. Values contained in a value set
+// must be unique.
+//
+// For Write Requests:
+//    INSERT will add the given values in the repeated field to the existing
+//    values in the value set.
+//    MODIFY will replace all existing values in the value set with the given
+//    values in the repeated field.
+//    DELETE will remove the values given in the repeated field from the value
+//    set if they exist, and will return error if given value(s) are not found.
+//    If only the value_set_id is specified (i.e. values_size() == 0), then all
+//    entries in the value set will be deleted.
+//
+// For Read Requests only the value_set_id should be specified and all the
+// entries in the value set should be returned in the read response.
+message ValueSetEntry {
+  uint32 value_set_id = 1;
+  repeated bytes values = 2;
```

Comment:
I went back to the old P4_14 spec to find some background on Value Sets.  I found this statement in section 4.3, "The run time API for updating parser value sets must allow value and mask pairs to be
specified together".  I don't see any place for a mask here, but I don't know whether that is relevant to this revival of value sets.

Reply:
In 7.13 of the PSA spec it suggests:
message ValueSetEntry {
        uint32 value_set_id = 1;
        // FieldMatch allows specification of exact, lpm, ternary, and
        // range matching on fields for tables, and these options are
        // permitted for the ValueSet extern as well.
        repeated FieldMatch match = 2;
 }

which is a wider set of values than what this pull request is proposing.  Is this discrepancy a problem.

Reply:
Thanks Tom for looking into this. I believe it should be ok to use repeated FieldMatch given @wmohsin suggestion to atomically apply write/read operations across all value set entries.
My only concern is the semantics of the `@size` annotation. Would a range match of 10:20 fit in a value set of size 1? Perhaps @hanw can confirm?

Reply:
+1 for @teverman's comment on using `FieldMatch` as per the PSA control-plane API proposal
Just like for a table, `size` refers to the logical size, not the actual number of physical entries required in HW.

Reply:
BTW, ValueSet will not be a PSA extern (at least in v1.0) but an experimental language feature. It should not have any bearing on the API though.

Reply:
@hanw I don't know what the most desirable answer is here, but some thoughts.

If by "logical size" you mean "the number of entries that the control plane wants to add, regardless of whether those entries are arbitrary ranges, exact match, lpm, or ternary entries", then yes, there are hardware implementations such as TCAMs where single range entries in the control plane turn into multiple entries in the hardware, so they are not one-for-one, and not easy to predict in the control plane what they will turn into in a P4 device, and that could differ from device to device, and based upon more criteria than merely whether a range is involved.

This is common between P4 tables and parser value sets.  So nothing really new here.  Whatever answer we want to give for what a 'size' attribute means, I suspect it should be the same for both.

Another thing that might be common to both tables and parser value sets - above a certain minimum size (say a few hundred entries), a table with exact match keys is often most efficiently implemented as a hash table, not a TCAM.  A TCAM implementation gives you nice predictable one-to-one capacity guarantees there between the control plane and data plane.  A hash table does not, not even if you use the common trick of supplementing it with a parallel "overflow TCAM" on the side.  The best you can do is specify some kind of notion of approximate size you want, e.g. I want this exact match table to support at least 800 entries.  A P4 compiler for that target might choose to use a 1024-entry hash table plus a small overflow TCAM for that.  Is it possible that a control plane could try to add 780 entries into this table, and one of them fails?  Unless you have some kind of magic (or you use a hardware TCAM), the answer is "yes, it might fail before you add 800 entries.  Can it be predicted in advance?  Not really, it is dependent on the contents of the particular keys, the hash function(s) in use".

Even more fun, suppose you successfully add 800 entries from the control plane, then reboot your system and the software forgets how those entries were packed into the hardware before.  If the control plane adds the _same_ 800 entries, but in a different order, some of them might fail the next time.

Again, unless you have some magic here, these are just realities of actual implementations, for which the best way to handle them that I know for hash tables are: "put in enough cushion so those failures are rare enough in practice that it doesn't matter for your use case".  Where what constitutes "rare enough" might depend on the use case at hand.

Fun?  Not really.  Reality?  Yes!

Reply:
Used logical notion of size (i.e. number of field matches). Seems the simplest to make progress on this.